ubcpi/ubcpi.py

```diff
@@ -0,0 +1,542 @@
+
+        """
+        if hasattr(key, 'to_deprecated_string'):
+            return key.to_deprecated_string()
+        else:
+            return unicode(key)
+
+
+@XBlock.needs('user')
+class PeerInstructionXBlock(XBlock, MissingDataFetcherMixin):
+    """
+    Peer Instruction XBlock
+
+    Notes:
+    storing index vs option text: when storing index, it is immune to the
+    option text changes. But when the order changes, the results will be
+    skewed. When storing option text, it is impossible (at least for now) to
+    update the existing students responses when option text changed.
+
+    a warning may be shown to the instructor that they may only do the minor
+    changes to the question options and may not change the order of the options,
+    add or delete options
+    """
+
+    # the display name that used on the interface
+    display_name = String(default="Peer Instruction")
+
+    question_text = Dict(
+        default={'text': 'What is the answer to life, the universe and everything?',
+                 'image_url': '', 'image_position': 'below', 'show_image_fields': 0, 'image_alt': ''},
+        scope=Scope.content,
+        help="The question the students see. This question appears above the possible answers which you set below. "
+             "You can use text, an image or a combination of both. If you wish to add an image to your question, press "
+             "the 'Add Image' button."
+    )
+
+    options = List(
+        # default=['Default Option 1', 'Default Option 2'], scope=Scope.content,
+        default=[
+            {'text': '21', 'image_url': '', 'image_position': 'below', 'show_image_fields': 0, 'image_alt': ''},
+            {'text': '42', 'image_url': '', 'image_position': 'below', 'show_image_fields': 0, 'image_alt': ''},
+            {'text': '63', 'image_url': '', 'image_position': 'below', 'show_image_fields': 0, 'image_alt': ''}
+        ],
+        scope=Scope.content,
+        help="The possible options from which the student may select",
+    )
+
+    rationale_size = Dict(
+        default={'min': 1, 'max': MAX_RATIONALE_SIZE}, scope=Scope.content,
+        help="The minimum and maximum number of characters a student is allowed for their rationale.",
+    )
+
+    correct_answer = Integer(
+        default=1, scope=Scope.content,
+        help="The correct option for the question",
+    )
+
+    correct_rationale = Dict(
+        default={'text': "Specify the rationale to explain the correct answer"}, scope=Scope.content,
+        help="The feedback for student for the correct answer",
+    )
+
+    stats = Dict(
+        default={'original': {}, 'revised': {}}, scope=Scope.user_state_summary,
+        help="Overall stats for the instructor",
+    )
+    seeds = List(
+        default=[], scope=Scope.content,
+        help="Instructor configured examples to give to students during the revise stage.",
+    )
+
+    sys_selected_answers = Dict(
+        default={}, scope=Scope.user_state_summary,
```

Comment:
We noticed this issue when developing. But I think this is the best place to store it so far. We haven't had many concrete ideas yet. we wouldn't expect to store a lot of info in this field. Maybe an additional weight or score for each answer.
I think this issue will exists for all xblocks that use this scope, including the example on xblock docs (http://xblock.readthedocs.org/en/ned-improved-docs/guide/xblock.html#fields). The longer the time that user interact with xblock before save, the higher possibility the data got messed up. 
I think giving xblock a little more control on when to load and save on the fields for this scope may help. So that the developer can delay the loading, write as early as possible and maybe put a transnational lock in between.